language/stdlib/src/transaction_scripts.rs

```diff
@@ -1,74 +1,216 @@
-    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
-    path.push(PathBuf::from(TRANSACTION_SCRIPTS));
-    path
+use std::{convert::TryFrom, env, fmt, fs, path::PathBuf};
+
+/// Cache mapping each script to its compiled bytes. This allows clients to get the bytes/hash of a
+/// script without invoking the compiler or reading from disk.
+static ALL_SCRIPTS: Lazy<Vec<(StdlibScript, CompiledBytes)>> = Lazy::new(|| {
+    StdlibScript::all()
+        .iter()
+        .map(|script| {
+            let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+            path.push(PathBuf::from(STAGED_OUTPUT_PATH));
+            path.push(PathBuf::from(TRANSACTION_SCRIPTS));
+            path.push(PathBuf::from(script.name()));
+            path.set_extension(STAGED_EXTENSION);
+            assert!(
+                path.exists(),
+                "Could not resolve path {} from working directory {}",
+                path.display(),
+                env::current_dir().unwrap().display()
+            );
+            let contents = fs::read(path).unwrap();
+            (*script, CompiledBytes(contents))
+        })
+        .collect()
 });
 
-/// We include the compiled transaction scripts as binaries since the docker files do not copy any
-/// source files over. This could then cause issues when the client wishes to use a pre-packaged
-/// transaction script.
+// This includes the compiled transaction scripts as binaries. We must use this hack to work around
+// a problem with Docker, which does not copy over the Move source files that would be be used to
+// produce these binaries at runtime.
+#[allow(dead_code)]
 const STAGED_TXN_SCRIPTS_DIR: Dir = include_dir!("staged/transaction_scripts");
 
-fn compiled_script(script_name: &str) -> Vec<u8> {
-    if use_staged() {
-        from_staged(script_name)
-    } else {
-        from_current_source(script_name)
-    }
+/// All of the Move transaction scripts that can be executed on the Libra blockchain
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub enum StdlibScript {
+    AddValidator,
+    CreateAccount,
+    EmptyScript,
+    Mint,
+    PeerToPeer,
+    PeerToPeerWithMetadata,
+    PlaceholderScript,
+    RegisterValidator,
+    RemoveValidator,
+    RotateAuthenticationKey,
+    RotateConsensusPubkey,
+    // ...add new scripts here
 }
 
-fn from_staged(script_name: &str) -> Vec<u8> {
-    let mut path = PathBuf::from(script_name);
-    path.set_extension(STAGED_EXTENSION);
-    STAGED_TXN_SCRIPTS_DIR
-        .get_file(path)
-        .unwrap()
-        .contents()
-        .to_vec()
-}
+impl StdlibScript {
+    // TODO: audit/cleanup. There is some cruft here; e.g., PlaceholderScript and EmptyScript are
+    // the same.
+    /// Return a vector containing all of the standard library scripts (i.e., all inhabitants of the
+    /// StdlibScript enum)
+    pub fn all() -> Vec<Self> {
+        vec![
+            Self::AddValidator,
+            Self::CreateAccount,
+            Self::EmptyScript,
```

Comment:
Should we contain the empty script in the whitelist? It wasn't there before.

Reply:
Discussed offline--we will keep everything under `transaction_scripts` in the whitelist for now and audit/simplify later (as the TODO above suggests).